SimplenoteTests/NoticeContorllerTests.swift

```diff
@@ -1,33 +1,171 @@
+            //No action
         }
+        let notice = Notice(message: "", action: action)
+
+        controller.present(notice)
+
+        let view = try XCTUnwrap(presenter.noticeView as? NoticeView)
+        XCTAssertEqual(notice.message, view.message)
+        XCTAssertNotEqual(notice.action?.title, view.actionTitle)
+        XCTAssertNotNil(view.handler)
     }
 
+    func testDismiss() {
+        let noticeA = Notice(message: "Message A", action: nil)
+
+        controller.present(noticeA)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+    }
+
+    func testDismissContinuesToNextNotice() {
+        let noticeA = Notice(message: "Message A", action: nil)
+        let noticeB = Notice(message: "Message B", action: nil)
+        controller.present(noticeA)
+        controller.present(noticeB)
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 1)
+
+        controller.dismiss()
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+        XCTAssertEqual(presenter.actionLog, "presentNoticeView(Message A), dismissNotification(), presentNoticeView(Message B), dismissNotification(), ")
+    }
+}
+
+extension NoticeContorllerTests {
+    static var timeInterval = TimeInterval.zero
+    static var timerNoActionCompletionHandler: (Timer) -> Void = { (_) in
+        //no action
+    }
+}
+
+class MockNoticePresenter: NoticePresenter {
+    var presented: Bool = false
+    var dismissed: Bool = false
+    var listeningToKeyboard: Bool = false
+    var presentedViews = [NoticeView]()
+    var actionLog = String()
+
+    override func presentNoticeView(_ noticeView: NoticeView, completion: @escaping () -> Void) {
+        self.noticeView = noticeView
+        presentedViews.append(noticeView)
+        actionLog.append("presentNoticeView(\(noticeView.message ?? "")), ")
+        presented = true
+
+        completion()
+    }
+
+    override func dismissNotification(completion: @escaping () -> Void) {
+        dismissed = true
+        actionLog.append("dismissNotification(), ")
+        completion()
+    }
+
+    override func startListeningToKeyboardNotifications() {
+        listeningToKeyboard = true
+    }
+}
+
+class MockTimerFactory: TimerFactory {
+    var timer: Timer?
+
+    override func scheduledTimer(with timeInterval: TimeInterval, completion: @escaping () -> Void) -> Timer {
+        if let timer = timer {
+            timer.fire()
+
+            return timer
+        } else {
+            return Timer.scheduledTimer(withTimeInterval: NoticeContorllerTests.timeInterval, repeats: false, block: NoticeContorllerTests.timerNoActionCompletionHandler)
+        }
+    }
 }
```

Comment:
What you want to do here is to return a mock timer that won't fire till you tell it to. This will allow you to precisely control test scenarios. Mock timer could look like this:
```
class MockTimerFactory: TimerFactory {
    var timer: Timer?

    override func scheduledTimer(with timeInterval: TimeInterval, completion: @escaping () -> Void) -> Timer {
        let timer = MockTimer()
        timer.completion = completion
        return timer
    }
}

class MockTimer: Timer {
    var completion: (() -> Void)?

    override func fire() {
        completion?()
    }

    override func invalidate() {
        super.invalidate()
        completion = nil
    }
}
```